records/providers_manager.go

```diff
@@ -0,0 +1,413 @@
+	// find the multiaddress associated with the returned peer id.
+	ProviderAddrTTL = amino.DefaultProviderAddrTTL
+)
+
+// ProvideValidity is the default time that a Provider Record should last on DHT
+// This value is also known as Provider Record Expiration Interval.
+var (
+	ProvideValidity        = amino.DefaultProvideValidity
+	defaultCleanupInterval = time.Hour
+	lruCacheSize           = 256
+	batchBufferSize        = 256
+	log                    = logging.Logger("providers")
+)
+
+// ProviderStore represents a store that associates peers and their addresses to keys.
+type ProviderStore interface {
+	AddProvider(ctx context.Context, key []byte, prov peer.AddrInfo) error
+	GetProviders(ctx context.Context, key []byte) ([]peer.AddrInfo, error)
+	io.Closer
+}
+
+// ProviderManager adds and pulls providers out of the datastore,
+// caching them in between
+type ProviderManager struct {
+	self peer.ID
+	// all non channel fields are meant to be accessed only within
+	// the run method
+	cache  lru.LRUCache
+	pstore peerstore.Peerstore
+	dstore *autobatch.Datastore
+
+	newprovs chan *addProv
+	getprovs chan *getProv
+
+	cleanupInterval time.Duration
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+var _ ProviderStore = (*ProviderManager)(nil)
+
+// Option is a function that sets a provider manager option.
+type Option func(*ProviderManager) error
+
+func (pm *ProviderManager) applyOptions(opts ...Option) error {
+	for i, opt := range opts {
+		if err := opt(pm); err != nil {
+			return fmt.Errorf("provider manager option %d failed: %s", i, err)
+		}
+	}
+	return nil
+}
+
+// CleanupInterval sets the time between GC runs.
+// Defaults to 1h.
+func CleanupInterval(d time.Duration) Option {
+	return func(pm *ProviderManager) error {
+		pm.cleanupInterval = d
+		return nil
+	}
+}
+
+// Cache sets the LRU cache implementation.
+// Defaults to a simple LRU cache.
+func Cache(c lru.LRUCache) Option {
+	return func(pm *ProviderManager) error {
+		pm.cache = c
+		return nil
+	}
+}
+
+type addProv struct {
+	ctx context.Context
+	key []byte
+	val peer.ID
+}
+
+type getProv struct {
+	ctx  context.Context
+	key  []byte
+	resp chan []peer.ID
+}
+
+// NewProviderManager constructor
+func NewProviderManager(ctx context.Context, local peer.ID, ps peerstore.Peerstore, dstore ds.Batching, opts ...Option) (*ProviderManager, error) {
+	pm := new(ProviderManager)
+	pm.self = local
+	pm.getprovs = make(chan *getProv)
+	pm.newprovs = make(chan *addProv)
+	pm.pstore = ps
+	pm.dstore = autobatch.NewAutoBatching(dstore, batchBufferSize)
+	cache, err := lru.NewLRU(lruCacheSize, nil)
+	if err != nil {
+		return nil, err
+	}
+	pm.cache = cache
+	pm.cleanupInterval = defaultCleanupInterval
+	if err := pm.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+	pm.ctx, pm.cancel = context.WithCancel(ctx)
+	pm.run()
+	return pm, nil
+}
+
+func (pm *ProviderManager) run() {
+	pm.wg.Add(1)
+	go func() {
+		defer pm.wg.Done()
```

Comment:
Since only one item is added to the WaitGroup, it would be preferable to use a channel instead.
```suggestion
	go func() {
		defer close(pm.closed)
```